cmd/notation/inspect.go

```diff
@@ -0,0 +1,288 @@
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runInspect(cmd, opts)
+		},
+	}
+
+	opts.LoggingFlagOpts.ApplyFlags(command.Flags())
+	opts.SecureFlagOpts.ApplyFlags(command.Flags())
+	cmd.SetPflagOutput(command.Flags(), &opts.outputFormat, cmd.PflagOutputUsage)
+	return command
+}
+
+func runInspect(command *cobra.Command, opts *inspectOpts) error {
+	// set log level
+	ctx := opts.LoggingFlagOpts.SetLoggerLevel(command.Context())
+
+	if opts.outputFormat != ioutil.OutputJson && opts.outputFormat != ioutil.OutputPlaintext {
+		return fmt.Errorf("unrecognized output format %s", opts.outputFormat)
+	}
+
+	// initialize
+	reference := opts.reference
+	sigRepo, err := getSignatureRepository(ctx, &opts.SecureFlagOpts, reference)
+	if err != nil {
+		return err
+	}
+
+	manifestDesc, ref, err := getManifestDescriptor(ctx, &opts.SecureFlagOpts, reference, sigRepo)
+	if err != nil {
+		return err
+	}
+
+	// reference is a digest reference
+	if err := ref.ValidateReferenceAsDigest(); err != nil {
+		if opts.outputFormat == ioutil.OutputPlaintext {
+			fmt.Printf("Resolved artifact tag `%s` to digest `%s` before inspect.\n", ref.Reference, manifestDesc.Digest.String())
+			fmt.Println("Warning: The resolved digest may not point to the same signed artifact, since tags are mutable.")
+		}
+		ref.Reference = manifestDesc.Digest.String()
+	}
+
+	if opts.outputFormat == ioutil.OutputPlaintext {
+		fmt.Println("Inspecting all signatures for signed artifact")
+		fmt.Println(ref.String())
+	}
+
+	output := inspectOutput{MediaType: manifestDesc.MediaType, Signatures: []signatureOutput{}}
+
+	err = sigRepo.ListSignatures(ctx, manifestDesc, func(signatureManifests []ocispec.Descriptor) error {
+		for _, sigManifestDesc := range signatureManifests {
+			sigBlob, sigDesc, err := sigRepo.FetchSignatureBlob(ctx, sigManifestDesc)
+			if err != nil {
+				return fmt.Errorf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %v", sigManifestDesc.Digest, manifestDesc.Digest, err.Error())
+			}
+
+			envelope, err := signature.ParseEnvelope(sigDesc.MediaType, sigBlob)
+			envelopeContent, err := envelope.Content()
+			if err != nil {
+				return err
+			}
+
+			signedArtifactDesc, err := notation.GetDescriptorFromPayload(&envelopeContent.Payload)
+			if err != nil {
+				return err
+			}
+
+			sig := signatureOutput{
+				Digest:                sigDesc.Digest.String(),
+				SignatureAlgorithm:    envelopeContent.SignerInfo.SignatureAlgorithm.String(),
+				SignedAttributes:      getSignedAttributes(envelopeContent),
+				UserDefinedAttributes: signedArtifactDesc.Annotations,
+				UnsignedAttributes:    getUnsignedAttributes(envelopeContent),
+				Certificates:          getCertificates(envelopeContent),
+				SignedArtifact: artifact{
+					MediaType: signedArtifactDesc.MediaType,
+					Digest:    signedArtifactDesc.Digest.String(),
+					Size:      signedArtifactDesc.Size,
```

Comment:
since the descriptor contains annotations, can't users add different user-metadata per signature?

Reply:
ah but the user-metadata would appear under user-defined attributes instead, so the the signed artifact here would be the same. in that case, yes, printing it multiple times seems redundant

Reply:
discussed separately, in the scenario that a signature for image A is attached to image B in a remote repository, it's possible for the signed artifact to be different across signatures despite being listed for the same reference. in that case, it makes sense to have the artifact displayed for all signatures

Reply:
> discussed separately, in the scenario that a signature for image A is attached to image B in a remote repository, it's possible for the signed artifact to be different across signatures despite being listed for the same reference. in that case, it makes sense to have the artifact displayed for all signatures

This is a very interesting point. Just curious, do we have any use case of attaching a signature of image A to image B? Because I think this signature would fail the verification process on verifying image B. So for the `inspect` command, do we want to show signatures that truly signed image B only?

Reply:
my impression was that we're just showing all signatures that are present in the sig repo for the reference, whether the payload truly matches or not. i am not sure about a specific use case around attaching signatures for different images, so i'd defer that decision to someone more familiar with the original motivation in the spec